Série de funcionalidades - POS e melhores

Shirkit opened the pull request:
Bem, venho trabalhando em atualização desde plugin desde o meio do ano passado, para a minha empresa. Nós usamos um sistema POS no WooCommerce, porém este plugin não tem suporte à NFC-e, apesar do WebMania ter a API pronta para tal.

Com isso em mente, comecei a atualizar o plugin segundo as minhas necessidades. Vou colocar um resumo de todas as atualizações que eu fiz.

Claro que eu não espero que esteja tudo ok de primeira, e imagino que devo algumas explicações, então vou colocar aqui em baixo parte do meu `functions.php` para ver para que servem os Hooks em questão.

Essa PR ainda está em DRAFT, pois gostaria de ouvir primeiro o que o @webmaniabr tem a dizer sobre a proposta.

Tem ainda trabalho há ser feito e ser revisado, mas acredito que a maior parte já foi feita. Especialmente, gostaria de identar todo o repositório segundo apenas um padrão, já que não é o caso aqui. Mas isso é pro final.

Vou deixar aqui por enquanto, ok?

**Sumário de mudanças**
```
Adicionado: opção de CNPJ da Credenciadora do cartão
Adicionado: botão para emitir NFC-e
Adicionado:  post_meta nfce
Adicionado: hooks diversos para controle e emissão de NFC-e e NF-e:
** Hooks obrigatórios **
 - `webmaniabr_modelo_nota` para controlar a emissão das notas fiscais
- `webmaniabr_pedido_pagamento` `webmaniabr_pedido_bandeira` `webmaniabr_pedido_autorizacao` não tentar adivinhar como o pagamento foi efetuado
Adicionado: endpoint que será usado pelo POS para realizar a chamada para emitir a NFC-e
Adicionado: emissão forçada de NF-e e NFC-e. Com hooks, é possível ignorar outras condições setadas anteriormente e forçar a emissão, com controle do desenvolvedor. Isso é para resolver problemas relacionados à NFC-e e emissão automática.
Adicionado: Bulk Edit (ainda não está 100%) para os campos de NF, para poder editar Classe de Imposto NCM etc.

** Hooks opcionais **
 - `webmaniabr_pedido_modalidade_frete` modificar possível valor padrão da modalidade do frete
- `webmaniabr_pedido_presenca` para indicar a presença do cliente
- `webmaniabr_pedido_tipo_integracao` indicação do tipo de integração do pedido
- `webmaniabr_pedido_valor_pagamento` o valor que foi pago na NFC-e. TODO: tem que pegar o valor padrão

Modificado: o método `order_data` foi modificado com adição de novos hooks para garantir a emissão de NFC-e
- `webmaniabr_pedido_cnpj_credenciadora` caso queira mudar o CNPJ da credenciadora para esta operação
Modificado: como a lista de gateways é pega, pois senão vai descartar gateways ocultos
Modificado: a tabela de Pedidos agora mostra diferença entre NF não emitidas, canceladas, anuladas, emitidas etc. Para maior controle do lojista.
Modificado: não permitir a reemissão de NF em caso de um pedido ter a NF não aprovada.
```

**functions.php**
```
<?php
/* ========================================================
 * Integrations with WebmaniaBR
 * ======================================================== */

// Fazer com que os pedidos do POS sejam vistos pelo WebmaniaBR
//add_action ('woocommerce_order_status_pending_to_completed', 'orquidario_woocommerce_order_status_pending_to_completed', 10, 1);
function orquidario_woocommerce_order_status_pending_to_completed($id)
{
   $order = wc_get_order($id);
   if ($order->get_created_via() == 'POS') {
	   if (class_exists('WooCommerceNFe')) {
		   WooCommerceNFe::instance()->emitirNFeAutomaticamenteOnStatusChange($id);
	   }
   }
}

// TODO: verificar quando, mesmo quando não solicitado pelo Cashier, se decide forçar a emissão ou não.
// NOTE: aqui vai dar override se a emissão automática estiver desabilitada
// NOTE: por exemplo, se for NFC-e e for no cartão, tem que emitir em TODOS os casos
add_filter('webmaniabr_emissao_automatica', 'orquidario_webmaniabr_emissao_automatica', 10, 3);
function orquidario_webmaniabr_emissao_automatica($force, $option, $post_id)
{
   if (!$force && get_post_type( $order_id ) == 'shop_order') {
	 $order = wc_get_order($post_id);
	 $register_id = get_post_meta($post_id, 'wc_pos_id_register', true);
	 $register = WC_Pos_Registers::instance()->get_register_name_by_id($register_id);
	 if (get_post_meta($post_id, 'wc_pos_order_type', true) !== '') {
		 $data = get_post_meta($post_id, 'card_payment_data', true);
		 if ($data)
		  return true;
	 }
   }
   return $force;
}

/*

Possíveis coisas:
1. Venda no orquidário - NFC-e - Presencial
2. Venda para entrega (aluguel ou outro) - NF-e - Não Presencial
3. Venda no shopping - NFC-e - Presencial fora
4. Venda direto pela internet
*/

// TODO: a condição para ser NFC-e tem que ser baseada na caixa que está sendo usado
// NOTE: se retornar algo sem ser 'nfe' ou 'nfce' não vai emitir a nota
add_filter('webmaniabr_modelo_nota', 'orquidario_webmaniabr_modelo_nota', 10, 2);
function orquidario_webmaniabr_modelo_nota($tipo, $post_id)
{
	$order = wc_get_order($post_id);
	$register_id = get_post_meta($post_id, 'wc_pos_id_register', true);
	$register = WC_Pos_Registers::instance()->get_register_name_by_id($register_id);
	if (get_post_meta($post_id, 'wc_pos_order_type', true) !== '') {
		$tipo = 'nfce';
	}
	return $tipo;
}

// TODO: a condição para a modalidade de frete é baseada em que? Se for NFC-e apenas?
//add_filter('webmaniabr_pedido_modalidade_frete', 'orquidario_webmaniabr_pedido_modalidade_frete', 10, 5);
function orquidario_webmaniabr_pedido_modalidade_frete($codigo, $modalidade, $modelo, $post_id, $order)
{
	if ((!$modalidade || $modalidade == 'null' || $modalidade = '') && $modelo == 2) {
		$codigo = 9;
	}
	return $codigo;
}

 // TODO: Na verdade tem que ver qual o caixa que está sendo usado.
add_filter('webmaniabr_pedido_presenca', 'orquidario_webmaniabr_pedido_presenca', 10, 4);
function orquidario_webmaniabr_pedido_presenca($presenca, $post_id, $modelo, $order)
{
	if ($order->get_user() === false && $modelo == 2 && get_post_meta($post_id, 'wc_pos_order_type', true) !== '') {
		$presenca = 1;
	}
	return $presenca;
}

// TODO: diferenciar à vista de à prazo?
//add_filter('webmaniabr_pedido_pagamento', 'orquidario_webmaniabr_pedido_pagamento', 10, 3 );
function orquidario_webmaniabr_pedido_pagamento($pagamento, $post_id, $order)
{
	return $pagamento;
}

//add_filter('webmaniabr_pedido_tipo_integracao', 'orquidario_webmaniabr_pedido_tipo_integracao' )
function orquidario_webmaniabr_pedido_tipo_integracao($tipo, $post_id, $order)
{
	return $tipo;
}

add_filter('webmaniabr_pedido_valor_pagamento', 'orquidario_webmaniabr_pedido_valor_pagamento', 10, 3);
function orquidario_webmaniabr_pedido_valor_pagamento($valor, $post_id, $order)
{
	$amt = get_post_meta($post_id, 'wc_pos_amount_pay', true);
	return $amt !== '' ? (float) $amt : $valor;
}

//add_filter('webmaniabr_pedido_cnpj_credenciadora', 'orquidario_webmaniabr_pedido_cnpj_credenciadora', 10, 3 );
function orquidario_webmaniabr_pedido_cnpj_credenciadora($cnpj, $post_id, $order)
{
	return $cnpj;
}

add_filter('webmaniabr_pedido_bandeira', 'orquidario_webmaniabr_pedido_bandeira', 10, 3);
function orquidario_webmaniabr_pedido_bandeira($bandeira, $post_id, $order)
{
	$data = get_post_meta($post_id, 'card_payment_data', true);
	if (isset($data['cardBrand'])) {
		if (stripos($data['cardBrand'], 'VISA') !== false) {
			return '01';
		} elseif (stripos($data['cardBrand'], 'MASTER') !== false) {
			return '02';
		} elseif (stripos($data['cardBrand'], 'ELECTRON') !== false) {
			return '01';
		} elseif (stripos($data['cardBrand'], 'MAESTRO') !== false) {
			return '02';
		} elseif (stripos($data['cardBrand'], 'ELO') !== false) {
			return '06';
		} elseif (stripos($data['cardBrand'], 'HIPERCARD') !== false) {
			return '07';
		} elseif (stripos($data['cardBrand'], 'AMEX') !== false) {
			return '03';
		} elseif (stripos($data['cardBrand'], 'AMERICAN') !== false) {
			return '03';
		} elseif (stripos($data['cardBrand'], 'SOROCRED') !== false) {
			return '04';
		} elseif (stripos($data['cardBrand'], 'CLUB') !== false) {
			return '05';
		} elseif (stripos($data['cardBrand'], 'DINER') !== false) {
			return '05';
		} elseif (stripos($data['cardBrand'], 'AURA') !== false) {
			return '08';
		} elseif (stripos($data['cardBrand'], 'CABAL') !== false) {
			return '09';
		} else {
			return '99';
		}
	}
	return $bandeira;
}

add_filter('webmaniabr_pedido_autorizacao', 'orquidario_webmaniabr_pedido_autorizacao', 10, 3);
function orquidario_webmaniabr_pedido_autorizacao($autorizacao, $post_id, $order)
{
	$data = get_post_meta($post_id, 'card_payment_data', true);
	if (isset($data['hostNsu'])) {
		return $data['hostNsu'];
	}
	return $autorizacao;
}

add_filter('option_wc_settings_woocommercenfe_payment_methods', 'orquidario_wc_settings_woocommercenfe_payment_methods', 10, 2);
function orquidario_wc_settings_woocommercenfe_payment_methods($value, $option)
{
	/*$value[get_option('pos_chip_pin_name')] = 03;
	$value[get_option('pos_chip_pin2_name')] = 04;
	$value[get_option('pos_chip_pin3_name')] = 99;*/
	$value['pos_chip_pin'] = 03;
	$value['pos_chip_pin2'] = 04;
	$value['pos_chip_pin3'] = 99;
	$value['cod'] = 01;
	return $value;
}
```

Shirkit committed:
Primeira parte

Ainda falta trabalho à ser feito, mas aqui é um bom princípio

Adicionado: opção de CNPJ da Credenciadora do cartão
Adicionado: botão para emitir NFC-e
Adicionado:  post_meta nfce
Adicionado: hooks diversos para controle e emissão de NFC-e e NF-e:
** Hooks obrigatórios **
 - `webmaniabr_modelo_nota` para controlar a emissão das notas fiscais
- `webmaniabr_pedido_pagamento` `webmaniabr_pedido_bandeira` `webmaniabr_pedido_autorizacao` não tentar adivinhar como o pagamento foi efetuado

** Hooks opcionais **
 - `webmaniabr_pedido_modalidade_frete` modificar possível valor padrão da modalidade do frete
- `webmaniabr_pedido_presenca` para indicar a presença do cliente
- `webmaniabr_pedido_tipo_integracao` indicação do tipo de integração do pedido
- `webmaniabr_pedido_valor_pagamento` o valor que foi pago na NFC-e. TODO: tem que pegar o valor padrão

Modificado: o método `order_data` foi modificado com adição de novos hooks para garantir a emissão de NFC-e
- `webmaniabr_pedido_cnpj_credenciadora` caso queira mudar o CNPJ da credenciadora para esta operação

Modificado: como a lista de gateways é pega, pois senão vai descartar gateways ocultos

Shirkit committed:
Update woocommerce_nfe.php

Removido partes inválidas

Shirkit committed:
Fix wrong variable

Shirkit committed:
Added API to request NF

Custom endpoint to request NF.

Shirkit committed:
Merge pull request #1 from webmaniabr/master

v 2.8.0

Shirkit committed:
Merge branch 'master' into dev

Shirkit committed:
Merge pull request #2 from Shirkit/dev

Merge do 2.8.2

Shirkit committed:
Fix for newest version

Shirkit committed:
Dont know why this was here

But maybe it's important?

Shirkit committed:
Merge branch 'master' of https://github.com/webmaniabr/NFeWooCommerce into webmaniabr-master

Shirkit committed:
Merge pull request #4 from Shirkit/webmaniabr-master

Update from Webmaniabr master - 2.9.11

Shirkit committed:
Merge branch 'dev'

Shirkit committed:
Merge branch 'master' of https://github.com/Shirkit/NFeWooCommerce

Shirkit committed:
Merge pull request #5 from Shirkit/master

Merge Oirigin update into working area

Shirkit committed:
Updated to lastest version from the origin repo

Shirkit committed:
Adicionado v1 de Bulk Edit para o NFE

Por enquanto só funcionam os campos texto (sem checkbox ou dropdown) e sem um estilo de verdade

Shirkit committed:
Corrigido a atualização de NF

Não faz sentido o `atualizar_ne` não sei quando seria setado isso. Será um bug?

Shirkit committed:
Permitir atualização de NFC-e

Salva corretamente quando for NFC-e, além de NF-e

Shirkit committed:
Corrigido impossível de se ver o status

Quando em status de processamento, já que a classe não se chama mais `processando`, e sim `processamento`

Shirkit committed:
Alterado tabela de pedidos

Agora a tabela de pedidos mostra se a nota está processando / contingência, e tem diferença se a nota não foi emitida ou se ela foi reprovada/cancelada.

E também fica amarelo nos casos, e vermelho quando deu erro.

Shirkit committed:
Não permitir re-emissão de notas

Quando já tem uma em processamento, aprovada, cancelada, reprovada ou em contingência

Shirkit committed:
Conflitos resolvidos

Shirkit committed:
Merge de DEV

Shirkit committed:
Corrigido código que sobrou de conflito

Shirkit committed:
Perrmitir gateways desativados

Cabe do administrador configurar, mas isso permite Gateways que são usdos em sistema POS que não são habilitados para a Frontend do WooCommerce apareçam pra este plugin.

webmaniabr commented:
Oi @Shirkit obrigado pelo pull request, vamos analisar todos os dados enviados e fazer os testes necessários. Pode demorar um pouco, estamos todos em homeoffice e com o atendimento reduzido. Assim que tiver novas atualizações pode ir colocando aqui também.

Abraço
Equipe WebmaniaBR®

Shirkit committed:
Permitir emissão para usuários "Visitantes"

Shirkit committed:
Permitir pessoa física como transportadora válida

Shirkit committed:
Fix: Uncaught Error: Call to undefined function menu_page_url()

Consertando erro causado quando salva as opções e `menu_page_url` não está disponível

Shirkit committed:
Corrigido emitirNFeAutomaticamenteOnStatusChange

Não é papel do `emitirNFeAutomaticamenteOnStatusChange` verificar o `status` do pedido ou se está ativo a emissão automática.

Afinal, ele é um hook que só é chamado quando registrado. Além disso, vai conflitar com a opção de customizar o `status` change do pedido.

Shirkit committed:
Movido o $force para outro lugar

TODO: tem que re-adicionar a opção para forçar a emissão por outro lugar, já que foi removido de `emitirNFeAutomaticamenteOnStatusChange`

Shirkit committed:
Acessando parâmetros de objetos em contexto estático

Shirkit committed:
Mais acessos fora de contexto

Shirkit committed:
Esqueci do $

Shirkit committed:
Permitir que notas sejam emitidas em status customizados

Shirkit committed:
Remoção de código já removido no fork original

Shirkit committed:
Permitir não emitir nota para determinado pedido

Shirkit committed:
Permitir alterar código NCM para variações

Shirkit committed:
Merge branch 'master' of https://github.com/webmaniabr/NFeWooCommerce into webmaniabr-master

Shirkit committed:
Merge branch 'webmaniabr-master'

Shirkit committed:
Corrigido erros de compilação

Shirkit committed:
Permitir escolher um tipo de unidade para cada produto

pa-philos commented:
Ainda adiciono funcionalidades, mas acredito que o PR já está maduro o suficiente para darem uma olhada.

Tem muita coisa diferente que foi trabalhada, em especial, a API REST que adicionei para os sistemas POS se comunicarem.

Ainda não está perfeito, mas acredito que já dê pra dar uma olhada.

Shirkit committed:
Corrigido bug da parte que pega o nome da pessoa para emissão da nota

Shirkit committed:
Ops, commit anterior estava errado

webmaniabr commented:
Olá @Shirkit tudo bem? Nós pedimos desculpas pela demora e queremos te agradecer por todas as sugestões de correções e melhorias. Publicamos a versão 3.1.0 do módulo com todos os Actions e Filters necessários para realizar a integração para emissão NFC-e no ponto de venda, sendo inspirado em suas sugestões.

Obrigado pela contribuição. Adicionamos o seu username no rodapé da Readme.

netwish-brasil commented:
Boa tarde, qual POS vocês usaram nesse exemplo "openpdv" ou "wcpos". Obrigado e parabéns pela iniciativa.

webmaniabr commented:
Oi @netwish-brasil, as modificações foram realizadas para atender a todos os plugins. Através dos hooks (actions and filters), você pode manipular todos os dados de emissão e assim processar nos plugins **openpdv** e **wcpos**.

Segue mais detalhes:

- https://ajuda.webmaniabr.com/hc/pt-br/articles/360052999091-Hooks-no-m%C3%B3dulo-de-Nota-Fiscal-WooCommerce-NFCe-e-Outros
- https://ajuda.webmaniabr.com/hc/pt-br/articles/360013064332-Como-realizar-integra%C3%A7%C3%A3o-da-NFC-e-com-meu-ponto-de-venda-
